Unify InlineLayout and BlockLayout (#659)
browserengineering/book: book/layout.md

each layout object, and finally draws each layout object to the
screen.

> [^no-box]: Elements like `<script>` don't generate layout objects, and

Reviewer: Isn't this still referenced on line 27?

class Layout:
    def __init__(self, node, parent, previous):
        self.node = node
        self.parent = parent
        self.previous = previous
        self.children = []
```

That said, requiring a `parent` and `previous` element now makes it
tricky to construct a `Layout` object in `Browser`, since the root of
the layout tree obviously can't have a parent. To rectify that, let me
add a second kind of layout object to serve as the root of the layout
tree.[^or-none] I think of that root as the document itself, so let's
call it `DocumentLayout`:

[^or-none]: I don't want to just pass `None` for the parent, because
the root layout object also computes its size and position
differently, as we'll see later this chapter.

``` {.python replace=%20Layout/%20BlockLayout dropline=display_list}
class DocumentLayout:
    def __init__(self, node):
        self.node = node
        self.parent = None
        self.children = []

    def layout(self):
        child = Layout(self.node, self, None)
        self.children.append(child)
        child.layout()
        self.display_list = child.display_list
```

Note an interesting thing about this new `layout` method: its role is

> to _create_ the child layout objects, and then _recursively_ calls

Reviewer: ```suggestion
to _create_ the child layout objects, and then _recursively_ call
```

which `self`, `previous`, and `next` point to. The two trees have
similar structure, so it's easy to get confused. But remember that
this code constructs the layout tree from the HTML tree, so it reads
from `node.children` (in the HTML tree) and writes to `self.children`
(in the layout tree).

So we have two ways to lay out an element: either calling `recurse`
and `flush`, or this `layout_intermediate` function. To determine
which one a layout object should to use, we'll need to know what kind
of content its HTML node contains: text and text-related tags like
`<b>`, or blocks like `<p>` and `<h1>`. That function looks something
like this:

``` {.python}
def layout_mode(node):
    if isinstance(node, Text):
        return "inline"
    elif node.children:
        if any([isinstance(child, Element) and \
                child.tag in BLOCK_ELEMENTS
                for child in node.children]):
            return "block"
        else:
            return "inline"
    else:
        return "block"
```

Here the list of `BLOCK_ELEMENTS` is basically what you expect, a list
of all the tags that describe parts of a page instead of
formatting:[^from-the-spec]

> [^from-the-spec]: Taken from the [HTML5 living standard][html5-elts].

Reviewer: ```suggestion
[^from-the-spec]: Taken from the [HTML living standard][html5-elts].
```

]
```

Our `layout_mode` function has to handle one tricky case, where a node

> contains both block objects like a `<p>` element but also text objects

Reviewer: ```suggestion
contains both block children like a `<p>` element but also text ones
```

to determine which layout mode to use for each element:
Finally, since `BlockLayout`s can now have children, the `layout`
method needs to recursively call `layout` so those children can
construct their children, and so on recursively:

``` {.python}
class BlockLayout:
    def layout(self):
        previous = None
        for child in self.node.children:
            if layout_mode(child) == "inline":
                next = InlineLayout(child, self, previous)
            else:
                next = BlockLayout(child, self, previous)
            self.children.append(next)
            previous = next
        # ...
        for child in self.children:
            child.layout()
```

Our layout tree now has a `DocumentLayout` at the root, `BlockLayout`s
at interior nodes, and `InlineLayout`s at the leaves:[^or-empty]
We also need to gather their `display_list` fields into a single
array:

[^or-empty]: Or, the leaf nodes could be `BlockLayout`s for empty
elements.
``` {.python expected=False}
class BlockLayout:
    def layout(self):
        # ...
        for child in self.children:
            self.display_list.extend(child.display_list)
```

> ::: {.widget big-height=490px small-height=860px}

Reviewer: Don't want this example any more?

Author: No, you're right, this one is still good (but I need to edit the caption). The other one (which had just `BlockLayout`s) is no longer good, since now the first stage is only inline, not only block.

]
```

Our `layout_mode` function has to handle one tricky case, where a node
contains both block children like a `<p>` element but also text
children like a text node or a `<b>` element. I've chosen to use block
mode in this case, but it's probably best to think of this as a kind
of error on the part of the web developer. And just like with implicit
tags in [Chapter 4](html.md), we use a repair mechanism to make sense
of the situation.[^anon-block]

[^anon-block]: In real browsers, that repair mechanism is called

> "[anonymous block boxes][anon-block]" and is more complex than what's

Reviewer: This link doesn't work in the rendered version.

the `layout_mode` of its HTML node:

``` {.python}
class BlockLayout:
    def layout(self):
        mode = layout_mode(self.node)
        if mode == "block":
            previous = None
            for child in self.node.children:
                # ...
        else:
            # ...
            self.recurse(self.node)
            self.flush()
```

Finally, since `BlockLayout`s can now have children, the `layout`
method needs to recursively call `layout` so those children can
construct their children, and so on recursively:

``` {.python}
class BlockLayout:
    def layout(self):
        # ...
        for child in self.children:
            child.layout()
```

We also need to gather their `display_list` fields into a single
array:

``` {.python expected=False}
class BlockLayout:
    def layout(self):
        # ...
        for child in self.children:
            self.display_list.extend(child.display_list)
```

Our browser is now constructing a whole tree of `BlockLayout` objects;
in fact, if you add a `print_tree` call to `Browser`'s `load` method,
you'll see that large web pages like this chapter produce large and
complex layout trees!

> ::: {.widget big-height=490px small-height=860px}

Reviewer: This widget is out of date - it references InlineLayout.

the `layout_mode` of its HTML node:

``` {.python}
class BlockLayout:
    def layout(self):
        mode = layout_mode(self.node)
        if mode == "block":
            previous = None
            for child in self.node.children:
                # ...
        else:
            # ...
            self.recurse(self.node)
            self.flush()
```

Finally, since `BlockLayout`s can now have children, the `layout`
method needs to recursively call `layout` so those children can
construct their children, and so on recursively:

``` {.python}
class BlockLayout:
    def layout(self):
        # ...
        for child in self.children:
            child.layout()
```

We also need to gather their `display_list` fields into a single
array:

``` {.python expected=False}
class BlockLayout:
    def layout(self):
        # ...
        for child in self.children:
            self.display_list.extend(child.display_list)
```

Our browser is now constructing a whole tree of `BlockLayout` objects;
in fact, if you add a `print_tree` call to `Browser`'s `load` method,
you'll see that large web pages like this chapter produce large and
complex layout trees!

> ::: {.widget big-height=490px small-height=860px}

Reviewer: The widget iframe is still not tall enough to avoid a vertical scrollbar.

Size and position
=================

In the [previous chapter](html.md), the `Layout` object was

> responsible for whole web page, so it just laid out its content

Reviewer: ```suggestion
responsible for the whole web page, so it just laid out its content
```

starting at the top of the page. Now that we have multiple
`BlockLayout` objects each containing a different paragraph of text,
we're going to need to do things a little differently, computing a
size and position for each layout object independently.

> Let's start with the `cursor_x` and `cursor_y`. Instead of having them

Reviewer: ```suggestion
Let's start with `cursor_x` and `cursor_y`. Instead of having them
```

be an important constraint in [later chapters](invalidation.md).

Now we'll need compute widths and heights. Width is easy: blocks are
as wide as their parents:[^until-css]

[^until-css]: In the [next chapter](styles.md), we'll add support for
author-defined styles, which in real browsers modify these layout
rules by setting custom widths or changing how *x* and *y* position
are computed.

``` {.python}
class BlockLayout:
    def layout(self):
        self.width = self.parent.width
        # ...
```

Height, meanwhile, is a little tricky. A `BlockLayout` that contains
other blocks should be tall enough to contain all of its children, so
its height should be the sum of its children's heights:

``` {.python}
class BlockLayout:
    def layout(self):
        # ...
        if mode == "block":
            self.height = sum([child.height for child in self.children])
```

However, a `BlockLayout` that contains text doesn't have children;
instead, it needs to be tall enough to contain all its text, which we
can conveniently read off of `cursor_y`:[^why-two-fields]

[^why-two-fields]: Since the height is just equal to `cursor_y`, why

>     not renamed `cursor_y` to `height` instead? You could, it would

Reviewer: ```suggestion
    not rename `cursor_y` to `height` instead? You could, it would
```

    work fine, but I would rather not. As you can see from, say, the
    `y` computation, the `height` field is a public field, read by
    other layout objects to compute their positions. As such I'd
    rather make sure it _always_ has the right value, whereas
    `cursor_y` changes as we lay out a paragraph of text and therefore
    sometimes has the "wrong" value. Keeping these two fields separate
    avoids a whole class of nasty bugs where the `height` field is
    read "too soon" and therefore gets the wrong value.

``` {.python}
class BlockLayout:
    def layout(self):
        # ...
        else:
            self.height = self.cursor_y
```

> Let's think again about dependencies. While the `x`, `y`, and `width`

Reviewer: Is this paragraph partially repeating the one starting at line 384?

`BlockLayout` itself; they now need to start from `0` instead of
`HSTEP` and `VSTEP`, both in `layout` and `flush`:

``` {.python}
class BlockLayout:
    def layout(self):
        else:
            self.cursor_x = 0
            self.cursor_y = 0

    def flush(self):
        # ...
        self.cursor_x = 0
        # ...
```

Since these fields are now relative, we'll need to add the block's `x`
and `y` position in `flush` when computing the display list:

``` {.python}
class BlockLayout:
    def flush(self):
        # ...
        for rel_x, word, font in self.line:
            x = self.x + rel_x
            y = self.y + baseline - font.metrics("ascent")
            self.display_list.append((x, y, word, font))
        # ...
```

Similarly, to wrap lines, we can't compare `cursor_x` to `WIDTH`,
because `cursor_x` is a relative measure while `WIDTH` is an absolute
measure; instead, we'll wrap lines when `cursor_x` reaches the block's
`width`:

``` {.python}
class BlockLayout:
    def text(self, node):
        for word in node.text.split():
            # ...
            if self.cursor_x + w > self.width:
                # ...
            # ...
```

> So now that sets us the problem of computing these `x`, `y`, and

Reviewer: ```suggestion
So now that leaves us with the problem of computing these `x`, `y`, and
```

`layout` call, so those children can compute their `x` and `y` based
on this block's `x` and `y`. Similarly, since the `y` position of a
block depends on its previous sibling's `y` position, the recursive
`layout` calls have to start and the first sibling and iterate through
the list forward---which is how we've already done it, but which will
be an important constraint in [later chapters](invalidation.md).

Now we'll need compute widths and heights. Width is easy: blocks are
as wide as their parents:[^until-css]

[^until-css]: In the [next chapter](styles.md), we'll add support for
author-defined styles, which in real browsers modify these layout
rules by setting custom widths or changing how *x* and *y* position
are computed.

``` {.python}
class BlockLayout:
    def layout(self):
        self.width = self.parent.width
        # ...
```

Height, meanwhile, is a little tricky. A `BlockLayout` that contains
other blocks should be tall enough to contain all of its children, so
its height should be the sum of its children's heights:

``` {.python}
class BlockLayout:
    def layout(self):
        # ...
        if mode == "block":

>             self.height = sum([child.height for child in self.children])

Reviewer: Needs more line wrapping to avoid a horizontal scrollbar.

                # ...
            # ...
```

So now that leaves us with the problem of computing these `x`, `y`,
and `width` fields. Let's recall that `BlockLayout`s represent blocks
of text like paragraphs or headings, and are stacked vertically one
atop another. That means each one starts at its parent's left edge:

``` {.python}
class BlockLayout:
    def layout(self):
        self.x = self.parent.x
        # ...
```

Its vertical position depends on the position and height of their
previous sibling. If there is no previous sibling, they start at the
parent's top edge:

``` {.python}
class BlockLayout:
    def layout(self):
        if self.previous:
            self.y = self.previous.y + self.previous.height
        else:
            self.y = self.parent.y
        # ...
```

Note that in each of these cases, to compute one block's `x` and `y`,
the `x` and `y` of its parent block must _already_ have been computed.
That means these computations have to go *before* the recursive
`layout` call, so those children can compute their `x` and `y` based
on this block's `x` and `y`. Similarly, since the `y` position of a
block depends on its previous sibling's `y` position, the recursive

> `layout` calls have to start and the first sibling and iterate through

Reviewer: ```suggestion
`layout` calls have to start at the first sibling and iterate through
```

For all three types of layout object, the order of the steps in the
`layout` method should be the same:

+ When `layout` is called, it first computes the `width`, `x`, and `y`
  fields, reading from the `parent` and `previous` layout objects.
+ Next, it creates a child layout object for each child element.
+ Then, the child layout nodes are recursively laid out by calling
  their `layout` methods.
+ Finally, `layout` computes the `height` field, reading from the
  child layout objects.

> You can see these steps in action in this widget:

Reviewer: This widget is broken in my local build on your branch.

For all three types of layout object, the order of the steps in the
`layout` method should be the same:

+ When `layout` is called, it first computes the `width`, `x`, and `y`
  fields, reading from the `parent` and `previous` layout objects.
+ Next, it creates a child layout object for each child element.
+ Then, the child layout nodes are recursively laid out by calling
  their `layout` methods.
+ Finally, `layout` computes the `height` field, reading from the
  child layout objects.

You can see these steps in action in this widget:

> 

Reviewer: Make this widget taller to avoid vertical scrolling.

            child.paint(display_list)
```

> You can now delete the line that computes a `BlockLayout`'s

Reviewer: Rephrase? Repetitive with the previous "you can now" sentence.
